src/FactorGraphs/Factor.php

```diff
@@ -70,44 +67,41 @@ protected function updateMessageVariable(Message $message, Variable $variable)
      */
     public function resetMarginals()
     {
-        $allValues = &$this->_messageToVariableBinding->getAllValues();
-        foreach ($allValues as &$currentVariable)
-        {
+        $allValues = $this->_messageToVariableBinding->getAllValues();
+        foreach ($allValues as $currentVariable) {
             $currentVariable->resetToPrior();
         }
     }
 
     /**
      * Sends the ith message to the marginal and returns the log-normalization constant
+     * @param $messageIndex
+     * @return
+     * @throws Exception
      */
     public function sendMessageIndex($messageIndex)
     {
         Guard::argumentIsValidIndex($messageIndex, count($this->_messages), "messageIndex");
 
-        $message = &$this->_messages[$messageIndex];
-        $variable = &$this->_messageToVariableBinding->getValue($message);
+        $message = $this->_messages[$messageIndex];
+        $variable = $this->_messageToVariableBinding->getValue($message);
         return $this->sendMessageVariable($message, $variable);
     }
 
-    protected abstract function sendMessageVariable(Message &$message, Variable &$variable);
+    protected abstract function sendMessageVariable(Message $message, Variable $variable);
 
-    public abstract function &createVariableToMessageBinding(Variable &$variable);
+    public abstract function createVariableToMessageBinding(Variable $variable);
 
-    protected function &createVariableToMessageBindingWithMessage(Variable &$variable, Message &$message)
+    protected function createVariableToMessageBindingWithMessage(Variable $variable, Message $message)
     {
-        $index = count($this->_messages);
-        $localMessages = &$this->_messages;
-        $localMessages[] = &$message;
         $this->_messageToVariableBinding->setValue($message, $variable);
-        $localVariables = &$this->_variables;
-        $localVariables[] = &$variable;
+        $this->_messages[] = $message;
+        $this->_variables[] = $variable;
         return $message;
     }
 
     public function __toString()
     {
-        return ($this->_name != null) ? $this->_name : base::__toString();
+        return ($this->_name != null) ? $this->_name : null;
```

Comment:
I mostly just changed it to null since it complained about base being an undefined constant (left overs since the port from C# I would assume) - the equivalent in PHP would be parent, but since we are in the uppermost class we can change it to $this->_name. Which is always set to a value in the constructor of Factor, so it should never be null. 

The only class that extends or uses ```Factor``` directly is ```GaussianFactor``` which in turn is used by
```
GaussianGreaterThanFactor
GaussianLikelihoodFactor
GaussianPriorFactor
GaussianWeightedSumFactor
GaussianWithinFactor
```

But all of these are setting their name via the parent constructor, and none of these are changing the name after they have been initialized. 

I actually forgot to revisit this after I've made the change to get this to work.